### DISCUSSION TITLE
Custom color opacity in theme function

### DISCUSSION THREAD

**Discussion by dawaltconley** [tailwindlabs/tailwindcss#8987] [Help] [answered]
Hi, I'm trying to reference custom colors with variable opacity elsewhere in my theme config and cannot figure out how. From the little bit of testing I've done it could be a bug or missing feature, but I wanted to ask here in case I'm missing something obvious.

Here are the relevant parts of my config:

```js
module.exports = {
  theme: {
    extend: {
      colors: {
        foobar: ({ opacityValue }) => `rgb(255 100 0 / ${opacityValue})`,
      },
      typography: ({ theme }) => ({
        theme: {
          css: {
            '--tw-prose-headings': theme('colors.red[500] / 50%'), // this works
            '--tw-prose-bullets': theme('colors.foobar / 80%'), // this does not
          },
        },
      }),
    },
  },
}
```

I've also tried using the new `<alpha-value>` syntax. The `text-foobar/80` etc classes _do_ work in my template files, but `opacityValue` seems to be `undefined` when referencing it in the config. i.e. the above outputs `rgb(255 100 0 / undefined)` for `--tw-prose-bullets`.

**Comment by wongjn** [comment, marked as answer]
Seems like it *might* be a bug (depending on whether it was ever intended that it should work for `color.foobar` and not just `color.foobar.shade`). To work-around this, set your color as a third-level item:

```diff
 module.exports = {
   theme: {
     extend: {
       colors: {
-        foobar: ({ opacityValue }) => `rgb(255 100 0 / ${opacityValue})`,
+        foobar: {
+          DEFAULT: 'rgb(255 100 0 / <alpha-value>)',
+        },
       },
       typography: ({ theme }) => ({
         theme: {
           css: {
             '--tw-prose-headings': theme('colors.red[500] / 50%'), // this works
-            '--tw-prose-bullets': theme('colors.foobar / 80%'), // this does not
+            '--tw-prose-bullets': theme('colors.foobar.DEFAULT / 80%'),
           },
         },
       }),
     },
   },
 }
```
I used the `DEFAULT` key so that it the color can still be referenced as `colors.foobar` elsewhere since `DEFAULT` is a special key in Tailwind which should cause little disruption in your existing codebase.


<details>
<summary>More in depth analysis of what's going on</summary>

The code that resolves alpha values in `theme()` function is:
https://github.com/tailwindlabs/tailwindcss/blob/14542d94f7078dc8d065ddcc6d403cd0a467f1be/src/util/resolveConfig.js#L176-L186

This expects there to be more that 2 components in the path; let's see why.

# 3-part color

## Initial state
```js
index = 0;
val = {
  // …
  colors: someResolvingFunction,
  // …
};
path = ['colors', 'red', '500']; // also path.alpha = '50%';
```

## While loop 1
```js
val = val[path[index++]]`
// index = 1; val = someResolvingFunction
let shouldResolveAsFn =
          isFunction(val) && (path.alpha === undefined || index < path.length - 1)
// shouldResolveAsFn = true, since index < path.length - 1 is true as index = 1, path.length - 1 = 2.
val = shouldResolveAsFn ? val(resolvePath, configUtils) : val
// val = {
//  …,
//  red: {
//    …,
//    500: '#ef4444',
//    …
//  },
//  …
//  }
```

## While loop 2
```js
val = val[path[index++]]`
// index = 2; val = val['red'] = { …, 500: '#ef4444', … }
let shouldResolveAsFn =
          isFunction(val) && (path.alpha === undefined || index < path.length - 1)
// shouldResolveAsFn = false, since isFunction(val) = false
val = shouldResolveAsFn ? val(resolvePath, configUtils) : val
// val = { …, 500: '#ef4444', … }
```

## While loop 3
```js
val = val[path[index++]]`
// index = 3; val = val['500'] = { …, 500: '#ef4444', … }['500'] = '#ef4444'
let shouldResolveAsFn =
          isFunction(val) && (path.alpha === undefined || index < path.length - 1)
// shouldResolveAsFn = false, since isFunction(val) = false
val = shouldResolveAsFn ? val(resolvePath, configUtils) : val
// val = '#ef4444'
```
While loop stops since the condition `index < path.length` now evaluates to `false` and our final `val` value is `#ef4444`, all good ✔.

# 2-part color

## Initial state
```js
index = 0;
val = {
  // …
  colors: someResolvingFunction,
  // …
};
path = ['colors', 'foobar']; // also path.alpha = '80%';
```

## While loop 1
```js
val = val[path[index++]]`
// index = 1; val = someResolvingFunction
let shouldResolveAsFn =
          isFunction(val) && (path.alpha === undefined || index < path.length - 1)
// shouldResolveAsFn = false, since index < path.length - 1 is false as index = 1, path.length - 1 = 1.
val = shouldResolveAsFn ? val(resolvePath, configUtils) : val
// val = someResolvingFunction
```

## While loop 2
```js
val = val[path[index++]]`
// index = 2; val = someResolvingFunction['foobar'] = undefined
// val is now undefined so no value is printed.
```
While loop stops since the condition `val !== undefined` now evaluates to `false` and our final `val` value is `undefined`, not good ❌.

# Summary

This is because in the first while loop of 2-part color, the `index < path.length - 1` condition stops `someResolvingFunction` from being evaluated as a function and instead passed through as-is.
</details>

**Reply by dawaltconley** [reply]
Thank you for the extremely thorough explanation! Using `DEFAULT` works as expected.
